Configurations/15-android.conf

```diff
@@ -0,0 +1,207 @@
+# set CROSS_SYSROOT [to $ANDROID_NDK/platforms/android-<api>/arch-<arch>]
+# and CROSS_COMPILE. CROSS_SYSROOT is still recognized [and even required
+# for some legacy targets], but if not set, it's detected and set to the
+# latest Android platform available with appointed NDK automatically. If
+# you need to target older platform, pass additional -D__ANDROID_API__=N
+# to Configure. For example, to compile for ICS on ARM with NDK 10d:
+#
+# ANDROID_NDK=/some/where/android-ndk-10d
+# PATH=$ANDROID_NDK/toolchains/arm-linux-androideabi-4.8/prebuild/linux-x86_64/bin:$PATH
+# [..]./Configure android-arm -D__ANDROID_API__=14
+#
+# One can engage clang by passing CC=clang to Configure. In such case
+# PATH needs even more adjustments to cover NDK's clang itself, as well
+# as unprefixed, yet target-specific ar and ranlib [or not, if you use
+# binutils-multiarch].
+
+{
+    my $android_ndk = {};
+    my %triplet = (
+        arm    => "arm-linux-androideabi",
+        arm64  => "aarch64-linux-android",
+        mips   => "mipsel-linux-android",
+        mips64 => "mips64el-linux-android",
+        x86    => "i686-linux-android",
+        x86_64 => "x86_64-linux-android",
+    );
+
+    sub android_ndk {
+        unless (%$android_ndk) {
+            my $ndk = $ENV{ANDROID_NDK};
+            die "\$ANDROID_NDK is not defined"  if (!$ndk);
+            die "\$ANDROID_NDK=$ndk is invalid" if (!-d "$ndk/platforms");
+
+            my $sysroot;
+
+            if (!($sysroot = $ENV{CROSS_SYSROOT})) {
+                my $api = "*";
+
+                # see if user passed -D__ANDROID_API__=N
+                foreach (@{$useradd{CPPDEFINES}}) {
+                    if (m|__ANDROID_API__=([0-9]+)|) {
+                        $api = $1;
+                        last;
+                    }
+                }
+
+                # list available platforms [numerically]
+                my @platforms = sort { $a =~ m/-([0-9]+)$/; my $aa = $1;
+                                       $b =~ m/-([0-9]+)$/; $aa <=> $1;
+                                     } glob("$ndk/platforms/android-$api");
+                die "no $ndk/platforms/android-$api" if ($#platforms < 0);
+
+                $config{target} =~ m|[^-]+-([^-]+)$|;   # split on dash
+                $sysroot = "@platforms[$#platforms]/arch-$1";
+            }
+            die "no sysroot=$sysroot"   if (!-d $sysroot);
+
+            $sysroot =~ m|/android-([0-9]+)/arch-(\w+)/?$|;
+            my ($api, $arch) = ($1, $2);
+
+            my $triarch = $triplet{$arch};
+            my $cflags = "-Wa,--noexecstack";
+            my $cppflags;
+
+            # see if user passed CC=clang
+            if ($user{CC} eq "clang") {
+                if (which("clang") !~ m|^$ndk/.*/prebuilt/([^/]+)/|) {
+                    die "no clang in \$ANDROID_NDK";
+                }
+                $cflags .= " -target $triarch -gcc-toolchain "
+                        . "\$(ANDROID_NDK)/toolchains/$triarch-4.9/prebuilt/$1";
```

Comment:
Target argument is not correct.
Let me quote snippet from shell script :
...
        case $ANDROID_ABI in
        armeabi)	LLVM_TRIPLET="${host_cpu}v5te"-none-$host_os;;
        armeabi-v7a*)	LLVM_TRIPLET="${host_cpu}v7"-none-$host_os;;
        *)		LLVM_TRIPLET=$host_cpu-none-$host_os;;
        esac
....
For example usualy is clang++ -target x86_64-none-linux-android  ..., i.e. -none is added after host cpu but with exceptions clang++ -target armv7-none-linux-androideabi ... and clang++ -target armv5te-none-linux-androideabi ...

Reply:
Thing is that $triarch is also reference to actual file names, so it has to be the way it is. Background check on how clang treats -target option has shown that suggested it's quite flexible. It effectively affects default and then it's perfectly possible to use -march. I've adjusted default so that it's same as for gcc. And since using gcc would also require -march flag, I'd say it's perfectly reasonable to expect that it would be required even in clang build.